From 5554d94a79a43e470a521a1b4fe4933b79d2bcb0 Mon Sep 17 00:00:00 2001
From: Richard Bradley <richard.bradley@joyent.com>
Date: Mon, 10 Jul 2017 10:18:19 +0000
Subject: [PATCH] MANATEE-341 "manatee-adm rebuild" hangs when rebuilding a
 sync peer MANATEE-344 Rebuild should not exit if postgres hasn't started on
 its first attempt

---
 lib/adm.js | 33 +++++++++++++++------------------
 1 file changed, 15 insertions(+), 18 deletions(-)

diff --git a/lib/adm.js b/lib/adm.js
index dd822ef..674d970 100644
--- a/lib/adm.js
+++ b/lib/adm.js
@@ -1507,11 +1507,20 @@ function rebuild(opts, cb) {
                     _addPostgresStatus,
                     function _getCurrentPeer(subctx, subcb) {
                         var thisPeer = null;
-                        subctx.state[_.shard].async.forEach(function (peer) {
-                            if (peer.zoneId == _.config.zoneId) {
-                                thisPeer = peer;
-                            }
-                        });
+
+                        if (subctx.state[_.shard].sync &&
+                            subctx.state[_.shard].sync.zoneId ===
+                            _.config.zoneId) {
+                            thisPeer = subctx.state[_.shard].sync;
+                        }
+                        if (subctx.state[_.shard].async && !thisPeer) {
+                            subctx.state[_.shard].async.forEach(function (p) {
+                                if (p.zoneId == _.config.zoneId) {
+                                    thisPeer = p;
+                                }
+                            });
+                        }
+
                         _.peer = thisPeer;
                         subcb();
                     },
@@ -1544,7 +1553,7 @@ function rebuild(opts, cb) {
                             process.stderr.write('\n');
                             if (restoreTry >= RESTORE_RETRIES) {
                                 return _cb(new Error('This Manatee instance ' +
-                                    'is not an active async after ' +
+                                    'is not an active peer after ' +
                                     RESTORE_RETRIES + ' restore attempts.  ' +
                                     'Check sitter logs.'));
                             }
@@ -1589,18 +1598,6 @@ function rebuild(opts, cb) {
                         }
 
                         lastRestore = restore;
-
-                    } else if (waitCount > 60) {
-                        /*
-                         * At this point no restore has taken place and
-                         * postgres hasn't started after roughly 60 seconds.
-                         * Something is likely wrong with postgres, so we error
-                         * here to prevent an unlimited spinner.
-                         */
-                        process.stderr.write('\n');
-                        return _cb(new Error('This Manatee instance is not ' +
-                            'active async after 60 seconds. Check sitter ' +
-                            'logs.'));
                     }
 
                     /*
-- 
2.21.0

